File: examples/ts/bounded-loop/BoundedLoop.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { TestContract } from 'runar-testing';

const __dirname = dirname(fileURLToPath(import.meta.url));
const source = readFileSync(join(__dirname, 'BoundedLoop.runar.ts'), 'utf8');

describe('BoundedLoop', () => {
  it('compiles via TestContract.fromSource', () => {
    const c = TestContract.fromSource(source, { expectedSum: 10n });
    expect(c.state.expectedSum).toBe(10n);
  });

  it('verify accepts the expected sum from start=0', () => {
    // sum_{i=0..4} (0+i) = 0+1+2+3+4 = 10
    const c = TestContract.fromSource(source, { expectedSum: 10n });
    const r = c.call('verify', { start: 0n });
    expect(r.success).toBe(true);
  });

  it('verify accepts the expected sum from start=2', () => {
    // sum_{i=0..4} (2+i) = 5*2 + 10 = 20
    const c = TestContract.fromSource(source, { expectedSum: 20n });
    const r = c.call('verify', { start: 2n });
    expect(r.success).toBe(true);
  });

  it('verify rejects mismatched expected sum', () => {
    const c = TestContract.fromSource(source, { expectedSum: 999n });
    const r = c.call('verify', { start: 0n });
    expect(r.success).toBe(false);
  });
});

File: examples/zig/bounded-loop/BoundedLoop_test.zig
const std = @import("std");
const root = @import("../examples_test.zig");
const BoundedLoop = @import("BoundedLoop.runar.zig").BoundedLoop;

fn contractPath(comptime basename: []const u8) []const u8 {
    return "bounded-loop/" ++ basename;
}

fn runCompileChecks(comptime basename: []const u8) !void {
    try root.runar.compileCheckSource(std.testing.allocator, @embedFile(basename), basename);
    try root.runar.compileCheckFile(std.testing.allocator, contractPath(basename));
}

test "compile-check BoundedLoop.runar.zig" {
    try runCompileChecks("BoundedLoop.runar.zig");
}

test "BoundedLoop.verify accepts the expected sum from start=0" {
    // Loop: 0+0 + 0+1 + 0+2 + 0+3 + 0+4 = 10.
    const c = BoundedLoop.init(10);
    c.verify(0);
}

test "BoundedLoop.verify accepts the expected sum from start=2" {
    // Loop: (2+0) + (2+1) + (2+2) + (2+3) + (2+4) = 5*2 + 10 = 20.
    const c = BoundedLoop.init(20);
    c.verify(2);
}
